lib/commons/aria/get-element-unallowed-roles.js

```diff
@@ -0,0 +1,85 @@
+/* global aria */
+/**
+ * gets all unallowed roles for a given node
+ * @method getElementUnallowedRoles
+ * @param {Object} node HTMLElement to validate
+ * @param {String} tagName tag name of a node
+ * @param {String} allowImplicit option to allow implicit roles, defaults to true
+ * @return {Array<String>} retruns an array of roles that are not allowed on the given node
+ */
+aria.getElementUnallowedRoles = function getElementUnallowedRoles(
+	node,
+	allowImplicit
+) {
+	/**
+	 * Get roles applied to a given node
+	 * @param {HTMLElement} node HTMLElement
+	 * @return {Array<String>} return an array of roles applied to the node, if no roles, return an empty array.
+	 */
+	// TODO: not moving this to outer namespace yet, work with wilco to see overlap with his PR(WIP) - aria.getRole
+	function getRoleSegments(node) {
+		let roles = [];
+		if (!node) {
+			return roles;
+		}
+		if (node.hasAttribute('role')) {
+			const nodeRoles = axe.utils.tokenList(
+				node.getAttribute('role').toLowerCase()
+			);
+			roles = roles.concat(nodeRoles);
+		}
+		if (node.hasAttributeNS('http://www.idpf.org/2007/ops', 'type')) {
+			const epubRoles = axe.utils
+				.tokenList(
+					node
+						.getAttributeNS('http://www.idpf.org/2007/ops', 'type')
+						.toLowerCase()
+				)
+				.map(role => `doc-${role}`);
+			roles = roles.concat(epubRoles);
+		}
+		return roles;
+	}
+
+	const tagName = node.nodeName.toUpperCase();
+
+	// by pass custom elements
+	if (!axe.utils.isHtmlElement(node)) {
+		return [];
+	}
+
+	// by pass other elements that may not be a known html element
+	if (
+		document.createElement(tagName).toString() === '[object HTMLUnknownElement]'
+	) {
+		return [];
+	}
+
+	const roleSegments = getRoleSegments(node);
+	const implicitRole = axe.commons.aria.implicitRole(node);
+	// stores all roles that are not allowed for a specific element most often an element only has one explicit role
+	return roleSegments.filter(role => {
```

Comment:
Why aren't you returning booleans in this?

Reply:
We have discussed this in the past, the result of this filter is an array of unallowed roles for a given element. 

This array is returned to `aria-allowed-role` check, which needs to populate the data for doT templating with an listing of roles for fail message. 

Hence we need to return role, than a boolean.

Reply:
Sorry Jey. I meant inside the callback. You're doing `return;` and `return role;` instead of `return false` / `return true`. It's nitpicky, I know. But returning undefined rather than false, and returning a string rather than true makes it look like this code is supposed to do something that it doesn't actually do.